tools/data_gen/src/main.rs

```diff
@@ -0,0 +1,197 @@
+use serde::ser::{Serialize, Serializer};
+use serde_json::{json, Map, Value};
+use std::borrow::Borrow;
+use std::cell::{Ref, RefCell};
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::rc::Rc;
+
+lazy_static! {
+    static ref SIZEMAP: HashMap<&'static str, Vec<&'static str>> = {
+        vec![
+            ("std_mem_d1", vec!["SIZE"]),
+            ("std_mem_d2", vec!["D0_SIZE", "D1_SIZE"]),
+            ("std_mem_d3", vec!["D0_SIZE", "D1_SIZE", "D2_SIZE"]),
+            (
+                "std_mem_d4",
+                vec!["D0_SIZE", "D1_SIZE", "D2_SIZE", "D3_SIZE"],
+            ),
+        ]
+        .into_iter()
+        .collect::<HashMap<&'static str, Vec<&'static str>>>()
+    };
+}
+
+///Holds data for std_mem cells, including name of cell, width, and sizes
+///Name is the name of cell itself, not its type. Sizes is a vector
+/// that holds the dimensions of the cell (ex: for a 2 x 3 std_mem_d2 cell it would be [2,3])
+struct CellData {
+    name: String,
+    width: u64,
+    sizes: Vec<u64>,
+}
+
+enum SizeVec {
+    Quad(Vec<Vec<Vec<Vec<u64>>>>),
+    Triple(Vec<Vec<Vec<u64>>>),
+    Double(Vec<Vec<u64>>),
+    Single(Vec<u64>),
+}
+
+impl Serialize for SizeVec {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match &*self {
+            SizeVec::Single(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 3, "data", &v)
+            }
+            SizeVec::Double(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 2, "data", &v)
+            }
+            SizeVec::Triple(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 1, "data", &v)
+            }
+            SizeVec::Quad(v) => {
+                serializer.serialize_newtype_variant("SizeVec", 0, "data", &v)
+            }
+        }
+    }
+}
+
+#[derive(FromArgs)]
+/// Used to read a file
+struct FilePath {
+    /// file path to read data from
+    #[argh(positional, from_str_fn(read_path))]
+    file: Option<PathBuf>,
+}
+
+fn read_path(path: &str) -> Result<PathBuf, String> {
+    Ok(Path::new(path).into())
+}
+
+fn main() -> CalyxResult<()> {
+    let fp: FilePath = argh::from_env();
+
+    let lib_path: &Path = &Path::new("../../../calyx");
+
+    let ws = frontend::Workspace::construct(&fp.file, lib_path)?;
+    let ctx: ir::Context = (ir::from_ast::ast_to_ir(ws))?;
+
+    let comp = ctx
+        .components
+        .into_iter()
+        .find(|comp| comp.name == ctx.entrypoint)
+        .expect("No top-level component found.");
+
+    let data_vec: Vec<CellData> = comp
+        .cells
+        .iter()
+        .filter_map(|cell| get_data(cell))
+        .collect();
+
+    let mut map = Map::new();
+
+    for CellData {
+        name,
+        width,
+        mut sizes,
+    } in data_vec
+    {
+        let json_comp: serde_json::Value = gen_comp(&mut sizes, width);
```

Comment:
The manual type annotation is unneeded here, but if you'd like to keep it, it can just be `Value` since you already imported `Value` from `serde_json` on line 7